src/agb_eeprom.c

```diff
@@ -0,0 +1,273 @@
+        // copy data into output buffer
+        for (t1 = 0; t1 < 4; t1++) {
+            value = 0;
+            for (t2 = 0; t2 < 0x10; t2++) {
+                value <<= 1;
+                value |= (*ptr++) & 1;
+            }
+            *(data--) = value;
+        }
+        return 0;
+    }
+}
+
+u16 EEPROMWrite1(u16 address, const u16* data) {
+    return EEPROMWrite(address, data, 1);
+}
+
+// reading from EEPROM like a status register
+#define REG_EEPROM (*(vu16*)0xdffff00)
+
+u16 EEPROMWrite(u16 address, const u16* data, u8 unk_3) {
+    u16 buffer[0x52]; // this is one too large?
+    vu16 timeout_flag;
+    vu16 prev_vcount;      // stack + a6
+    vu16 current_vcount;   // stack + a8
+    vu32 passed_scanlines; // stack + ac
+    u16 ret;
+    vu16* temp2;
+
+    u32 r2;
+
+    u8 i, j;
+    u16* ptr;
+
+    if (address >= gEEPROMConfig->size)
+        return EEPROM_OUT_OF_RANGE;
+
+    ptr = (u16*)(0x42 + (uintptr_t)&buffer + (uintptr_t)(gEEPROMConfig->address_width * 2) + 0x42);
+    *ptr-- = 0;
+    // copy data into buffer
+    for (i = 0; i < 4; i++) {
+        r2 = *data++;
+        for (j = 0; j < 16; j++) {
+            *ptr = 1&r2;
+            ptr--;
+            r2 = r2 >> 1;
+        }
+    }
+
+    // copy address to buffer
+    for (i = 0; i < gEEPROMConfig->address_width; i++) {
+        *ptr = 1&address;
+        ptr--;
+        address = address >> 1;
+    }
+    *ptr-- = 0;
+    *ptr-- = 1;
+    DMA3Transfer(buffer, (u16*)0xdffff00, gEEPROMConfig->address_width + 0x43);
+    ret = 0;
+    timeout_flag = 0;
+    prev_vcount = REG_VCOUNT;
+    passed_scanlines = 0;
+
+    while (1) {
+        if (!timeout_flag) {
+            if (REG_EEPROM & 1) {
+                timeout_flag++;
+                if (!unk_3)
+                    break;
+            }
+        }
+
+        current_vcount = REG_VCOUNT;
+        if (current_vcount != prev_vcount) {
+            if (current_vcount > prev_vcount) {
+                passed_scanlines += (current_vcount - prev_vcount);
+            } else {
+                passed_scanlines += (current_vcount - (prev_vcount - 0xE4));
+            }
+
+            if (passed_scanlines > 0x88) {
+                if (timeout_flag)
+                    break;
+                if ((REG_EEPROM & 1)) {
+                    break;
+                }
+
+                ret = 0xc001;
+                break;
+            }
+            prev_vcount = current_vcount;
+        }
+    }
+
+    return ret;
+}
+
+u16 EEPROMCompare(u16 address, const u16* data) {
+    u16 ret;
+
+    u16 buffer[4];
+    u16* ptr;
+
+    u8 i;
+
+    ret = 0;
+    if (address >= gEEPROMConfig->size) {
+        return EEPROM_OUT_OF_RANGE;
+    }
+    EEPROMRead(address, buffer);
+    ptr = buffer;
+    for (i = 0; i < ARRAY_COUNT(buffer); i++) {
+        if (*data++ != *ptr++) {
+            ret = EEPROM_COMPARE_FAILED;
+            break;
+        }
+    }
+    return ret;
+}
+
+const char EEPROM_NOWAIT[] = "EEPROM_NOWAIT";
```

Comment:
Not linked properly and not following the convention. global vars startswith `g`, static vars startswith `s` (but gUnk... is acceptable since it's going to be renamed some time).